### PR TITLE
Stamping Mechanism Refactoring

### PR DESCRIPTION
**Pull request by Nusnus**
W.I.P at the moment

### CODE REVIEW

**Comment by auvipy** [comment]
splitting to smaller chunks only or more improvement on the existing implementation?

**Comment by auvipy** [comment]
thank you for your in detail response

File: celery/canvas.py [hunk 1]
```diff
@@ -162,33 +165,33 @@ def on_signature(self, sig, **headers) -> dict:
              Dict: headers to update.
          """
 
-    def on_chord_header_start(self, chord, **header) -> dict:
```
Comment (Nusnus) [inline celery/canvas.py:165]: This caused a very annoying bug where `headers` had a key named `chord` and this caused an exception to be raised.

File: t/unit/tasks/test_canvas.py [hunk 2]
```diff
@@ -1019,15 +585,11 @@ def s(*args, **kwargs):
         assert c.tasks[-1].options['chord'] == 'some_chord_id'
 
         c.apply_async(link=[s(32)])
-        expected_sig = s(32)
-        expected_sig.stamp(visitor=GroupStampingVisitor())
```
Comment (Nusnus) [inline t/unit/tasks/test_canvas.py:1023]: This was an API breaking change the `GroupStampingVisitor` introduced that could have broken prod systems post 5.3 release...

File: t/integration/test_stamping.py [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+            elif i >= check_max:
+                assert count == expected_count
+        # try again later
+        sleep(check_interval)
+    else:
+        raise TimeoutError(f"{redis_key!r} was never incremented")
+
+    # There should be no more increments - block momentarily
+    sleep(min(1, timeout))
+    assert int(redis_connection.get(redis_key)) == expected_count
```
Comment (thedrow) [inline t/integration/test_stamping.py:115]: I suggest that we share this code via a common module. I don't like the code duplication if there's no reason for it.

**Review by thedrow — CHANGES_REQUESTED** [review]

File: celery/canvas.py [hunk 4]
```diff
@@ -582,22 +553,47 @@ def stamp(self, visitor=None, **headers):
...
+                ])
+            ]
+            if "stamped_headers" in headers:
+                headers["stamped_headers"].extend(stamped_headers)
+            else:
+                headers["stamped_headers"] = stamped_headers
+            for stamp in headers.keys():
+                if stamp != "stamped_headers" and stamp not in headers["stamped_headers"]:
+                    headers["stamped_headers"].append(stamp)
+            headers["stamped_headers"] = list(set(headers["stamped_headers"]))
```
[the remark this answers is not in the document]
Comment (Nusnus) [reply celery/canvas.py]: I agree. Thank's for the code snippet.
Fix pushed.

File: celery/canvas.py [hunk 5]
```diff
@@ -582,22 +553,47 @@ def stamp(self, visitor=None, **headers):
...
+            _merge_dictionaries(headers, self.options, aggregate_duplicates=False)
 
         # Preserve previous stamped headers
         stamped_headers = set(self.options.get("stamped_headers", []))
-        stamped_headers.update(headers["stamped_headers"])
+        stamped_headers.update(headers.get("stamped_headers", []))
         headers["stamped_headers"] = list(stamped_headers)
```
[the remark this answers is not in the document]
Comment (Nusnus) [reply celery/canvas.py:593]: `# Preserve previous stamped headers`
We merged the visitor headers + headers args. This adds the headers from `self`:
`stamped_headers = set(self.options.get("stamped_headers", []))`

File: celery/canvas.py [hunk 6]
```diff
@@ -582,22 +553,47 @@ def stamp(self, visitor=None, **headers):
...
-        return self.set(**headers)
+        for previous_header in stamped_headers:
+            if previous_header not in headers and previous_header in self.options:
+                headers[previous_header] = self.options[previous_header]
+        return headers
```
[the remark this answers is not in the document]
Comment (Nusnus) [reply celery/canvas.py:597]: I'm not sure I understand your comment, can you please explain 🙏
Also, please notice the diff between `stamp()` and `_stamp_headers()`, the github diff view might make it look confusing, because I'm not fully sure why you say:
> So now we set all headers every time we stamp? Even the non-stamped headers?

And what difference do you see from the previous implementation logic-wise?

File: t/integration/test_stamping.py (hunk 3 above)
Comment (Nusnus) [reply t/integration/test_stamping.py]: @thedrow I did not work at all in this PR on the integration tests besides moving it to an external file, so I will address this only after I finish working on the integration tests.
For now I need to close open ends before the integration tests at all so you'll have to review it again when I'll be done 🙏

File: celery/canvas.py [hunk 7]
```diff
@@ -582,22 +553,47 @@ def stamp(self, visitor=None, **headers):
             headers (Dict): Stamps that should be added to headers.
         """
         self.stamp_links(visitor, **headers)
-
-        headers = headers.copy()
+        headers = deepcopy(headers)
```
[the remark this answers is not in the document]
Comment (Nusnus) [reply celery/canvas.py]: Removed

File: celery/canvas.py [hunk 8]
```diff
@@ -582,22 +553,47 @@ def stamp(self, visitor=None, **headers):
...
+        visitor_headers = None
         if visitor is not None:
-            visitor_headers = visitor.on_signature(self, **headers)
+            visitor_headers = visitor.on_signature(self, **headers) or {}
+        headers = self._stamp_headers(visitor_headers, **headers)
+        return self.set(**headers)
+
+    def _stamp_headers(self, visitor_headers=None, **headers):
+        headers = deepcopy(headers)
```
[the remark this answers is not in the document]
Comment (Nusnus) [reply celery/canvas.py]: Removed

File: celery/canvas.py [hunk 9]
```diff
@@ -608,40 +604,37 @@ def stamp_links(self, visitor, **headers):
...
         """
-        if not visitor:
-            return
-
-        non_visitor_headers = headers.copy()
+        non_visitor_headers = deepcopy(headers)
 
         # Stamp all of the callbacks of this signature
-        headers = non_visitor_headers.copy()
+        headers = deepcopy(non_visitor_headers)
```
[the remark this answers is not in the document]
Comment (Nusnus) [reply celery/canvas.py:611]: TBH I don't have a clear explanation.
If I use `.copy()` the unit tests are failing (some).
If I use `deepcopy()`, it works as expected.

I've replaced all other instances with `.copy()`, but here with `non_visitor_headers` we do need `deepcopy()` or the feature breaks.

I'm not sure why but there are much bigger issues at hand, so I didn't invest a lot investigating this.

File: celery/app/task.py [hunk 10]
```diff
@@ -953,41 +953,7 @@ def replace(self, sig):
...
-                    links.extend(task.options['link_error'] if 'link_error' in task.options else [])
-
-            for link in links:
-                link_stamped_headers = stamped_headers.copy()
-                link_stamped_headers.extend(link['options'].get('stamped_headers', []))
-                link_stamped_headers = list(set(link_stamped_headers))
-                link['options']['stamped_headers'] = link_stamped_headers
-                link['options'].update(stamps)
-
+        self.on_replace_stamping(sig)
```
[the remark this answers is not in the document]
Comment (Nusnus) [reply celery/app/task.py:956]: Sure np.
Please notice Sphinx doc will be the last step after everything is working and the code is cleaned and fixed 🙂

[the remark this answers is not in the document]
Comment (Nusnus) [reply celery/app/task.py]: Fix pushed

File: celery/canvas.py [hunk 11]
```diff
@@ -70,7 +70,7 @@ def _stamp_regen_task(task, visitor, **headers):
     return task
 
 
-def _merge_dictionaries(d1, d2):
+def _merge_dictionaries(d1, d2, aggregate_duplicates=True):
```
[the remark this answers is not in the document]
Comment (Nusnus) [reply celery/canvas.py:73]: Fix pushed

> Can you please explain what we use aggregate_duplicates for?

Read both (new) doc in `_merge_dictionaries` and in `_stamp_headers`

File: celery/canvas.py [hunk 12]
```diff
@@ -582,22 +553,47 @@ def stamp(self, visitor=None, **headers):
...
+        return self.set(**headers)
+
+    def _stamp_headers(self, visitor_headers=None, **headers):
+        headers = deepcopy(headers)
+        if visitor_headers is not None:
+            visitor_headers = visitor_headers or {}
             if "stamped_headers" not in visitor_headers:
                 visitor_headers["stamped_headers"] = list(visitor_headers.keys())
-            _merge_dictionaries(headers, visitor_headers)
+            _merge_dictionaries(headers, visitor_headers, aggregate_duplicates=False)
```
[the remark this answers is not in the document]
Comment (Nusnus) [reply celery/canvas.py]: Fix pushed

File: celery/canvas.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Nusnus) [reply celery/canvas.py]: Fix pushed

File: celery/canvas.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Nusnus) [reply celery/canvas.py:624]: Please review again @thedrow , I've refactored the code to fix the previous comments so it might be easier to understand now

File: celery/canvas.py [hunk 13]
```diff
@@ -582,22 +553,47 @@ def stamp(self, visitor=None, **headers):
...
+            headers["stamped_headers"] = list(set(headers["stamped_headers"]))
         else:
-            headers["stamped_headers"] = [header for header in headers.keys() if header not in self.options]
-            _merge_dictionaries(headers, self.options)
+            stamped_headers = [
+                header for header in headers.keys()
+                if all([
+                    header != "stamped_headers",
+                    header in self.options.get("stamped_headers", [])
+                ])
```
[the remark this answers is not in the document]
Comment (Nusnus) [reply celery/canvas.py]: This was removed.
Please review again @thedrow

File: celery/canvas.py [hunk 14]
```diff
@@ -582,22 +553,47 @@ def stamp(self, visitor=None, **headers):
...
+            ]
+            if "stamped_headers" in headers:
+                headers["stamped_headers"].extend(stamped_headers)
+            else:
+                headers["stamped_headers"] = stamped_headers
+            for stamp in headers.keys():
```
[the remark this answers is not in the document]
Comment (Nusnus) [reply celery/canvas.py]: I'm not sure what comment you want for `result_extended`.
Can you please review it again (the code was fixed and changed a lot) and suggest a comment to add? 🙏

Thanks!

File: celery/canvas.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Nusnus) [reply celery/canvas.py]: In the last version of this code it was completely removed so please review it again and ignore:
> I agree. Thank's for the code snippet.
> Fix pushed.

@thedrow

**Comment by Nusnus** [comment]
@thedrow I've replied to all of your comments.
All of the required fixes are pushed.
Please review again and notice:
1. The integration tests are basically untouched from `main` at this point, only moved to a different file so far into the PR.
2. Sphinx documentation will be the last step, to make sure we have a stable version first.

File: celery/canvas.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Nusnus) [reply celery/canvas.py:624]: > I guess your comment should explain what happens during replacing that requires us to always set all stamped headers even if they were not touched by the visitor or by the kwargs.

TBH I don't know. The new test cases are checking so many new edge cases that in the case of stamping with task replacement, this happened.

It is not perfect but it will require a lot of time investigating it just for the comment whereas the feature is still not completed, so if it's acceptable, I'd prefer to leave this code with the comment referencing the replace, but not investigate more at the moment as its too expensive for the sake of having a more detailed comment.

**Comment by auvipy** [comment]
> @auvipy @thedrow Do you have any idea why does the code coverage is failing?
> 
> > Failing after 1s — 86.99% (-0.01%) compared to [cb83eaf](https://github.com/celery/celery/commit/cb83eafe59782963cb0254edce23f485f03bfbf1)

that's mostly random codcov issue I guess

**Comment by Nusnus** [comment]
> > @auvipy @thedrow Do you have any idea why does the code coverage is failing?
> > > Failing after 1s — 86.99% (-0.01%) compared to [cb83eaf](https://github.com/celery/celery/commit/cb83eafe59782963cb0254edce23f485f03bfbf1)
> 
> that's mostly random codcov issue I guess

I think I found the reason it failed. Very interesting!
Fixed in this commit: https://github.com/celery/celery/pull/8045/commits/db7936b6c45b62e81e7db8022030ce79ac152d84

The code that I removed in this commit was code that indeed ran, but the condition `if stamp in self.options and stamp not in link.options:` was always false after all of the fixes and changes of this PR.
So technically, the body of the condition was never accessed, thus causing a code coverage error.

Now the interesting part - investigating this, I discovered that the original code (the one I just removed in https://github.com/celery/celery/pull/8045/commits/db7936b6c45b62e81e7db8022030ce79ac152d84) was actually a patch to a bug in the stamping mechanism, which while the bug existed this code was accessed so it did not have any code cov errors when it was first introduced. As this PR fixes the stamping mechanism, the patch fixes an impossible edge case now (as it is naturally handled now, without need for edge cases patches), so the workaround wasn't needed anymore, making the code coverage error in the PR actually very useful in finding out useless code that can be removed safely.

It was damn hard to understand but finally I can get my 100% passing CI, I hate seeing errors! 😁

@auvipy @thedrow 
I think this makes a very important lesson - we shouldn't ignore very easily CI errors, even if it's "just code coverage".

**Review by auvipy — APPROVED** [review]
looks good

**Review by thedrow — APPROVED** [review]

**Comment by Nusnus** [comment]
Thank you!! 🙏 🙏
